scylla/tests/ccm_integration/ccm/logged_cmd.rs

```diff
@@ -0,0 +1,318 @@
+use anyhow::{Context, Error};
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::io;
+use std::ops::Deref;
+use std::process::{ExitStatus, Stdio};
+use std::sync::atomic::AtomicI32;
+use tokio::io::{AsyncBufReadExt, BufReader};
+use tokio::process::Command;
+
+/// A simple abstraction to run commands, and emit logs during its execution.
+/// It should allow to run multiple commands in parallel.
+pub(crate) struct LoggedCmd {
+    run_id: AtomicI32,
+}
+
+/// A set of options for the command to run.
+pub(crate) struct RunOptions {
+    /// Environment variables for the command.
+    env: HashMap<String, String>,
+    /// A flag telling whether the command is allowed to fail.
+    /// If set to true, and command fails, the error is not propagated.
+    allow_failure: bool,
+}
+
+impl RunOptions {
+    /// The default run options. With empty environment and `allow_failure` set to false.
+    pub(crate) fn new() -> Self {
+        RunOptions {
+            env: HashMap::new(),
+            allow_failure: false,
+        }
+    }
+
+    pub(crate) fn with_env(mut self, env: HashMap<String, String>) -> Self {
+        self.env = env;
+        self
+    }
+
+    pub(crate) fn allow_failure(mut self, allow: bool) -> Self {
+        self.allow_failure = allow;
+        self
+    }
+}
+
+impl LoggedCmd {
+    pub(crate) async fn new() -> Self {
+        LoggedCmd {
+            run_id: AtomicI32::new(1),
+        }
+    }
+
+    async fn process_child_result(
+        status: io::Result<ExitStatus>,
+        allow_failure: bool,
+        run_id: i32,
+        stderr: Vec<u8>,
+        command_with_args: String,
+    ) -> Result<ExitStatus, Error> {
+        match status {
+            Ok(status) => {
+                match status.code() {
+                    Some(code) => {
+                        tracing::info!("{:15} -> status = {}", format!("exited[{}]", run_id), code);
+                    }
+                    None => {
+                        tracing::info!("{:15} -> status = unknown", format!("exited[{}]", run_id));
+                    }
+                }
+                if !allow_failure && !status.success() {
+                    let tmp = stderr.deref();
+                    return Err(Error::msg(format!(
+                        "Command `{}` failed: {}, stderr: \n{}",
+                        command_with_args,
+                        status,
+                        std::str::from_utf8(tmp)?
+                    )));
```

Comment:
:question: About usage of `?`: do we really want to propagate the `str::from_utf8` conversion error instead of returning the error that the command failed?